### ISSUE TITLE
`yarn npm audit` incorrectly reports security advisories do not exist under certain circumstances

### ISSUE THREAD

**Issue by davidmally-at** [yarnpkg/berry#3778]
I believe I've discovered a bug in the way **yarn 3** audits for security advisories using `yarn npm audit`. If Package X is installed at Version A, and this version has a security advisory reported for it, `yarn npm audit` will correctly report the advisory. However, if you also install Package Y, which itself installs Version B of Package X that is not affected by the security advisory, `yarn npm audit` will no longer report the advisory if Version B is higher than Version A. Presumably this does not happen if Version B is lower than Version A, but I have not tested it. While `yarn npm audit` reports that the security advisory is no longer there, Version A of the package is still being used, and the vulnerable code is still being included.

See [this repository](https://github.com/davidmally-at/yarn-npm-audit-bug) for a PoC of the bug :)

I'm a security engineer, and this tool is a core part of my workflow for tracking and patching vulnerabilities, and it would be awesome if this were fixed so I can keep using and trusting yarn.

**Comment by Novolokov** [comment]
Appears this is still an issue on 3.6.1. 

```
$ yarn npm audit --all
➤ YN0001: No audit suggestions
```

```
$ yarn why immer
├─ @reduxjs/toolkit@npm:1.9.0
│  └─ immer@npm:9.0.21 (via npm:^9.0.16)
│
├─ @reduxjs/toolkit@npm:1.9.0 [eede8]
│  └─ immer@npm:9.0.21 (via npm:^9.0.16)
│
├─ pre-front-end@workspace:.
│  └─ immer@npm:9.0.21 (via npm:^9.0.12)
│
├─ react-dev-utils@npm:12.0.1
│  └─ immer@npm:9.0.21 (via npm:^9.0.7)
│
└─ react-dev-utils@npm:12.0.1 [53cd1]
   └─ immer@npm:9.0.21 (via npm:^9.0.7)
```

```
$ yarn --version
3.6.1

$ npm --version
9.8.1
```

**Comment by merceyz** [comment]
Indeed, the PR that fixes this was a breaking change so it will be in v4.
You can use one of the release candidates if you need it before the final release.
```sh
yarn set version canary
```

**Comment by ethanluginbuhl** [comment]
Hi, I'm still having the problem discussed in this issue. I'm new to yarn. What should I do?

**Comment by thefonso** [comment]
Anyone from the security community able to verify if this is resolved as of v4.6.0? cc: @davidmally-at
